java/java.editor/src/org/netbeans/modules/editor/java/JavaCodeTemplateFilter.java

```diff
@@ -72,88 +73,125 @@ private JavaCodeTemplateFilter(JTextComponent component, int offset) {
+                                            }
+                                            String headerText = controller.getText().substring(startPos, so);
+                                            int idx = headerText.indexOf('{'); //NOI18N
+                                            if (idx < 0) {
+                                                treeKindCtx = null;
+                                                stringCtx = CLASS_HEADER;
+                                            }
+                                            break;
+                                        case FOR_LOOP:
+                                        case ENHANCED_FOR_LOOP:
+                                            if (!isRightParenthesisOfLoopPresent(component, controller)) {
+                                                 treeKindCtx = null;
+                                            }
+                                            break;
+                                        case PARENTHESIZED:
+                                            if (isPartOfWhileLoop(component, controller)) {
+                                                if (!isRightParenthesisOfLoopPresent(component, controller)) {
+                                                        treeKindCtx = null;
                                                 }
-                                        }
+                                            }
+                                            break;
                                     }
                                 }
-                            });
-                        } catch (ParseException ex) {
-                            Exceptions.printStackTrace(ex);
-                        }
+                            }
+                        });
+                    } catch (ParseException ex) {
+                        Exceptions.printStackTrace(ex);
                     }
                 }, NbBundle.getMessage(JavaCodeTemplateProcessor.class, "JCT-init"), cancel, false); //NOI18N
             }
         }
     }
+    
+    private boolean isPartOfWhileLoop(JTextComponent component, CompilationController controller) {
+        TreeUtilities treeUtilities = controller.getTreeUtilities();
+        TreePath currentPath = treeUtilities.pathFor(component.getCaretPosition());
+        TreePath parentPath = treeUtilities.getPathElementOfKind(Tree.Kind.WHILE_LOOP, currentPath);
+        return parentPath != null;
+    }
+    
+    private boolean isRightParenthesisOfLoopPresent(JTextComponent component, CompilationController controller) {
+        AtomicBoolean result = new AtomicBoolean(true);
+        Document document = component.getDocument();
+        document.render(() -> {
+            TokenHierarchy<?> tokenHierarchy = controller.getTokenHierarchy();
+            TokenSequence<?> tokenSequence = tokenHierarchy.tokenSequence();
+            tokenSequence.move(component.getCaretPosition());
```

Comment:
As far as I can tell from the line `String text = controller.getText().substring(startPos, so);` the wrong value of the end index of the substring is used, namely the start offset of the abbreviation, although (the end offset of the abbreviation + the number of spaces before the closing bracket, if any + 1) should be used to check for the presence of the right parenthesis (the problem is that the end offset of the abbreviation is not calculated before, at least in the  JavaCodeTemplateFilter class). Because of this, the `treeKindCtx` variable is always set to `null` and the code template for the for loop is rejected. 

In addition, the code for the `WHILE_LOOP` case is never executed, because when the caret position is in parentheses that are part of a while loop, the leaf node for the tree path on which the caret is located is `ParenthesizedTree`, not `WhileLoopTree`. Because of this, the template for a while loop works even when there is no right parenthesis.

Reply:
@jlahoda Done as suggested. Please review.